bftengine/src/bftengine/PersistentStorageImp.cpp

```diff
@@ -317,112 +315,104 @@ void PersistentStorageImp::initDescriptorOfLastExecution() {
 void PersistentStorageImp::setDescriptorOfLastExecution(const DescriptorOfLastExecution &desc) {
   setDescriptorOfLastExecution(desc, false);
   hasDescriptorOfLastExecution_ = true;
+  descriptorOfLastExecution_ = DescriptorOfLastExecution{desc.executedSeqNum, desc.validRequests};
 }
 
 /***** Windows handling *****/
 
 /***** Private functions *****/
 
-void PersistentStorageImp::serializeAndSaveSeqNumWindow(const SharedPtrSeqNumWindow &seqNumWindow) {
-  writeBeginningOfActiveWindow(BEGINNING_OF_SEQ_NUM_WINDOW, seqNumWindow.get()->getBeginningOfActiveWindow());
-  for (uint32_t i = seqNumWindowFirst_; i < seqNumWindowLast_; ++i)
-    setSeqNumDataElement(i, seqNumWindow);
+void PersistentStorageImp::saveDefaultsInSeqNumWindow() {
+  writeBeginningOfActiveWindow(BEGINNING_OF_SEQ_NUM_WINDOW, seqNumWindowBeginning_);
+  const SeqNumData seqNumData;
+  for (uint32_t i = 0; i < seqWinSize; ++i)
+    setSeqNumDataElement(i, seqNumData);
 }
 
 void PersistentStorageImp::setSeqNumDataElement(const SeqNum &index, const SeqNumData &seqNumData) const {
-  char *buf = new char[SeqNumData::maxSize()];
-  SeqNum shift = (index - 1) * numOfSeqNumWinParameters;
-  char *movablePtr = buf;
+  UniquePtrToChar buf(new char[SeqNumData::maxSize()]);
+  SeqNum shift = index * numOfSeqNumWinParameters;
+  char *movablePtr = buf.get();
   size_t actualSize = seqNumData.serializePrePrepareMsg(movablePtr);
   uint32_t itemId = BEGINNING_OF_SEQ_NUM_WINDOW + PRE_PREPARE_MSG + shift;
   Assert(itemId < BEGINNING_OF_CHECK_WINDOW);
-  metadataStorage_->writeInTransaction(itemId, buf, actualSize);
+  metadataStorage_->writeInTransaction(itemId, buf.get(), actualSize);
   Assert(actualSize != 0);
 
-  movablePtr = buf;
+  movablePtr = buf.get();
   actualSize = seqNumData.serializeFullCommitProofMsg(movablePtr);
   itemId = BEGINNING_OF_SEQ_NUM_WINDOW + FULL_COMMIT_PROOF_MSG + shift;
   Assert(itemId < BEGINNING_OF_CHECK_WINDOW);
-  metadataStorage_->writeInTransaction(itemId, buf, actualSize);
+  metadataStorage_->writeInTransaction(itemId, buf.get(), actualSize);
   Assert(actualSize != 0);
 
-  movablePtr = buf;
+  movablePtr = buf.get();
   actualSize = seqNumData.serializePrepareFullMsg(movablePtr);
```

Comment:
I just realized something with regards to these `movablePtr` use cases. The only reason we need this pattern at all is because the functions like `serializePrepareFullMsg` take a reference to a pointer. Therefore they change the pointer passed in, rather than just copying it and moving the copy in the parameter. We can get rid of all these movablePtrs if we take `char* buf` instead of `char*& buf` as a parameter. We are literally fighting ourselves because of references to pointers that we don't actually want. In the future, we should probably be suspicious of any code that takes a `*&` parameter. It's not really a necessary thing to do in most c++ code.

Reply:
That's the way how ifstream works. It moves the pointer to after serialized data position. I have to support two separate cases: serialization of singe parameters and of whole elements containing a bunch of parameters. In the second case, this functionality is required.

Reply:
Ok. I really think there's gotta be a cleaner way than aliasing pointers, but let's defer this for now.